### PR TITLE
Upgrade the code base to support type declarations features (PHP 7.4 and prior)

### PR DESCRIPTION
**Pull request by AmraniCh**
## CHANGES

* As the title says, this PR concerns upgrading our code base to support type declaration features that are added in PHP version `7.4` and prior.

## EXPLANATION

JoliTypo support only `>=7.4` PHP versions, so we can take advantage of all PHP `<=7.4` features to upgrade our library code base, this PR focused mainly on the integration of the following features (others features can be integrated later):

* [Typed Properties](https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.typed-properties) (*>= v7.4*).
* [Scalar Type Declarations](https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.scalar-type-declarations) (*>= v7.0*).
* [Return type declarations](https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.return-type-declarations) (*>= v7.0*).

### BREAKING CHANGE

This is a breaking change, if this PR is approved and planned to get merged, I suggest doing so in the next major version of JoliTypo (version `2.0.0`), because if these changes get merged and a non-breaking version is released `1.x.x`, and one of our library users has their own implementation of our fixer interface (`FixerInterface`) or just they extends our classes and they override parent methods, then they will get some PHP fatal errors, this is a simple example that demonstrates this scenario:

```php
interface FixerInterface
{
    public function fix(string $content, ?StateBag $stateBag = null): string;
}

/**
 * Example of a fixer that extends the fixer interface
 */
class CustomFixer implements FixerInterface
{
    public function fix(string $content, ?StateBag $stateBag = null)
    {
        //
    }
}
```

This code will trigger this error:

```
PHP Fatal error:  Declaration of CustomFixer::fix(string $content, ?JoliTypo\StateBag $stateBag = null) must be compatible with FixerInterface::fix(string $content, ?JoliTypo\StateBag $stateBag = null): string in ...
```

### STRICT MODE

By adding declaration types we actually do not strict our properties and function arguments to their declared types, instead, PHP will try to coerce values of wrong types into the closed expected type declaration. however strict typing is disabled by default and it can be enabled at runtime by doing `declare(strict_types=1);` on a per-file basis, if we planned to enable strict typing I suggest enabling strict typing in a different PR after merging this one if approved of course.

## CONCLUSION

If this PR is approved and merged I can work on other PRs to integrate the other `7.4` features and have our code base fully use PHP `<=7.4` features.

### CODE REVIEW

File: src/JoliTypo/Bridge/Symfony/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -43,7 +43,10 @@ public function getConfigTreeBuilder(): TreeBuilder
         return $treeBuilder;
     }
 
-    private function getRootNode(TreeBuilder $treeBuilder, $name)
+    /**
+     * @return \Symfony\Component\Config\Definition\Builder\NodeDefinition|\Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition
```
Comment (lyrixx) [inline src/JoliTypo/Bridge/Symfony/DependencyInjection/Configuration.php:47]: Can you import these symbol?

```
use Symfony\Component\Config\Definition\Builder\NodeDefinition;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
```
...

File: src/JoliTypo/Bridge/Twig/JoliTypoExtension.php [hunk 2]
```diff
@@ -28,14 +28,18 @@ public function getFunctions(): array
         ];
     }
 
+
```
Comment (lyrixx) [inline src/JoliTypo/Bridge/Twig/JoliTypoExtension.php:31]: oups :)

File: src/JoliTypo/Bridge/Twig/JoliTypoExtension.php [hunk 3]
```diff
@@ -28,14 +28,18 @@ public function getFunctions(): array
...
     public function getFilters(): array
     {
         return [
             new \Twig\TwigFilter('jolitypo', [$this, 'translate'], ['pre_escape' => 'html', 'is_safe' => ['html']]),
         ];
     }
 
-    public function translate($text, $preset = 'default')
+    /**
+     * @throws \Symfony\Component\Config\Definition\Exception\InvalidConfigurationException
```
Comment (lyrixx) [inline src/JoliTypo/Bridge/Twig/JoliTypoExtension.php:40]: same comment for the import. Can you fix them all?. Thanks

File: src/JoliTypo/Bridge/Twig/JoliTypoExtension.php [hunk 4]
```diff
@@ -44,7 +48,7 @@ public function translate($text, $preset = 'default')
         return $this->presets[$preset]->fix($text);
     }
 
-    public function getName()
+    public function getName(): string
```
Comment (lyrixx) [inline src/JoliTypo/Bridge/Twig/JoliTypoExtension.php:51]: This method is not used anymore since twig ... I don't remember, but long enough to be removed :) 

(ok, I [searched](https://github.com/twigphp/Twig/blob/1.x/CHANGELOG#L333-L339)...  6 years)

Can you removed it at all?

File: src/JoliTypo/Fixer.php [hunk 5]
```diff
@@ -46,27 +46,24 @@ class Fixer
     /**
      * @var array HTML Tags to bypass
      */
-    protected $protectedTags = ['head', 'link', 'pre', 'code', 'script', 'style'];
+    protected array $protectedTags = ['head', 'link', 'pre', 'code', 'script', 'style'];
```
Comment (lyrixx) [inline src/JoliTypo/Fixer.php:49]: Unfortunatelly, this is a BC break here too. And IMHO,this is part of the public API.

https://3v4l.org/ZkNVb

@damienalexandre are you OK with this BC break?

File: src/JoliTypo/Fixer/BaseOpenClosePair.php [hunk 6]
```diff
@@ -17,12 +17,12 @@
...
 abstract class BaseOpenClosePair
 {
-    protected function fixViaState($content, StateBag $stateBag, $stateName, $openRegexp, $closeRegexp, $openReplacement, $closeReplacement)
+    protected function fixViaState(string $content, StateBag $stateBag, string $stateName, string $openRegexp, string $closeRegexp, string $openReplacement, string $closeReplacement): string
     {
         $storedSibling = $stateBag->getSiblingNode($stateName);
 
         // If no stored open quote node & open quote detected
-        if (false === $storedSibling && preg_match($openRegexp, $content)) {
+        if (null === $storedSibling && preg_match($openRegexp, $content)) {
```
Comment (lyrixx) [inline src/JoliTypo/Fixer/BaseOpenClosePair.php:25]: Why this change?

File: src/JoliTypo/Fixer/SmartQuotes.php [hunk 7]
```diff
@@ -17,17 +17,17 @@
 
 class SmartQuotes extends BaseOpenClosePair implements FixerInterface, LocaleAwareFixerInterface
 {
-    protected $opening;
-    protected $openingSuffix = '';
-    protected $closing;
-    protected $closingPrefix = '';
+    protected string $opening       = '';
```
Comment (lyrixx) [inline src/JoliTypo/Fixer/SmartQuotes.php:20]: same here

File: src/JoliTypo/FixerInterface.php [hunk 8]
```diff
@@ -12,10 +12,8 @@
 interface FixerInterface
 {
     /**
-     * @param string   $content  A string to fix
-     * @param StateBag $stateBag A bag of useful information
-     *
-     * @return string
+     * @param string        $content  A string to fix
+     * @param StateBag|null $stateBag A bag of useful information
```
Comment (lyrixx) [inline src/JoliTypo/FixerInterface.php:16]: Can be removed, it doesn't bring more information :)

File: src/JoliTypo/FixerInterface.php [hunk 9]
```diff
@@ -12,10 +12,8 @@
...
      */
-    public function fix($content, StateBag $stateBag = null);
+    public function fix(string $content, ?StateBag $stateBag = null): string;
```
Comment (lyrixx) [inline src/JoliTypo/FixerInterface.php:18]: the return type must be removed here, since this is the very most important part of the public API

File: src/JoliTypo/LocaleAwareFixerInterface.php [hunk 10]
```diff
@@ -11,8 +11,5 @@
 
 interface LocaleAwareFixerInterface
 {
-    /**
-     * @param string $locale
-     */
-    public function setLocale($locale);
+    public function setLocale(string $locale): void;
```
Comment (lyrixx) [inline src/JoliTypo/LocaleAwareFixerInterface.php:14]: same here

File: src/JoliTypo/StateBag.php [hunk 11]
```diff
@@ -11,48 +11,32 @@
 
 class StateBag
 {
-    /**
-     * @var int
-     */
-    protected $currentDepth = 0;
+    protected int $currentDepth = 0;
```
Comment (lyrixx) [inline src/JoliTypo/StateBag.php:14]: IMHO, this class is internal, and it not extended (as it's useless anyway).

So could you

1. turn every property to private
2. add `@internal` on top of it?

**Review by lyrixx — COMMENTED** [review]
## Thanks

You did a terrific job by opening the PR. It's super cool to see new contributor
on this project. We would like to thank you for your effort !

## About the BC

We don't want to break the public API.

But adding return type to private things are OK. Moreover we'll allow to break
the BC for some minor things, like internal stuff. For example, you can add
return type to:

* src/JoliTypo/Bridge/Twig/JoliTypoExtension.php
* src/JoliTypo/StateNode.php

but not:

* src/JoliTypo/Fixer.php
* src/JoliTypo/FixerInterface.php

If someone might extends a class from this lib, it must continue to goes well.

But for very internal stuff (`private`, or not meant to be extended) it's OK.

BTW, can you also add `@internal` to the class that are not meant to be extended?


## About strict mode

strict mode in PHP does not work well, and it's not a good idea to use it. It
brings a feeling of robustness, but it's not the case. We prefer to not use it.
(That's why we don't use it in Symfony neither)

---

Thanks again for your contribution. We'll be happy to merge it once you'll fix
the BC break.

File: src/JoliTypo/Bridge/Symfony/DependencyInjection/Configuration.php (hunk 1 above)
Comment (AmraniCh) [reply src/JoliTypo/Bridge/Symfony/DependencyInjection/Configuration.php]: done!

File: src/JoliTypo/Bridge/Twig/JoliTypoExtension.php (hunk 2 above)
Comment (AmraniCh) [reply src/JoliTypo/Bridge/Twig/JoliTypoExtension.php]: done!

File: src/JoliTypo/Bridge/Twig/JoliTypoExtension.php (hunk 3 above)
Comment (AmraniCh) [reply src/JoliTypo/Bridge/Twig/JoliTypoExtension.php]: done!

File: src/JoliTypo/Bridge/Twig/JoliTypoExtension.php (hunk 4 above)
Comment (AmraniCh) [reply src/JoliTypo/Bridge/Twig/JoliTypoExtension.php:50]: Emm, okay but for things to be organized and for easy tracking and documenting changes, I suggest removing it in a separate PR (I will create it after this PR gets merged), what do you think ?

File: src/JoliTypo/Fixer.php (hunk 5 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer.php:49]: Yes, it is a breaking change, I suggest removing all types of declarations from protected properties and methods.

Comment (damienalexandre) [reply src/JoliTypo/Fixer.php:49]: Just did a quick and not very scientific lookup:

https://github.com/search?q=protected+%24protectedTags+Fixer&type=Code

And I consider this list internal - never seen anyone changing the default value, considering there is also a setter if someones needs to change the value.

https://github.com/jolicode/JoliTypo/blob/d85b1076486b6886f1a16943d032d11b9c06c78d/src/JoliTypo/Fixer.php#L131

Marking it as private is ok for me.

File: src/JoliTypo/FixerInterface.php (hunk 8 above)
Comment (AmraniCh) [reply src/JoliTypo/FixerInterface.php:16]: Emm this is true if we decided to keep the method argument type declarations, but as you said before `Fixer` and `FixerInterface` should not be affected by this PR changes.

File: src/JoliTypo/Fixer.php (hunk 5 above)
Comment (lyrixx) [reply src/JoliTypo/Fixer.php:49]: @damienalexandre my question also involves all other protected properties :)

Comment (damienalexandre) [reply src/JoliTypo/Fixer.php:49]: All the properties that have a setter could be changed to **private** IMO.

File: src/JoliTypo/FixerInterface.php (hunk 8 above)
Comment (AmraniCh) [reply src/JoliTypo/FixerInterface.php]: done!

File: src/JoliTypo/FixerInterface.php (hunk 9 above)
Comment (AmraniCh) [reply src/JoliTypo/FixerInterface.php]: done!

File: src/JoliTypo/LocaleAwareFixerInterface.php (hunk 10 above)
Comment (AmraniCh) [reply src/JoliTypo/LocaleAwareFixerInterface.php]: done!

File: src/JoliTypo/Fixer.php (hunk 5 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer.php:49]: I suggest doing this in a separate PR that I will create after merging this one.

> All the properties that have a setter could be changed to **private** IMO.

For now, I will remove all return type declarations from methods and properties that are part of the public API:

* `FixerInterface` and `LocaleAwareFixerInterface`.
* All the fixer classes `JoliTypo\Fixer\*` including the abstract class `JoliTypo\Fixer\BaseOpenClosePair`.
* `StateBag`.

This is okey ?

File: src/JoliTypo/Fixer/BaseOpenClosePair.php (hunk 6 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer/BaseOpenClosePair.php:25]: That is because the `$storeSibling` could be either a `StateNode` instance or `NULL`, I changed the false value in `StateBag::getSiblingNode` to NULL to support the return type declaration `?StateNode`.

https://3v4l.org/sTkPF#v7.4.33

File: src/JoliTypo/Fixer/SmartQuotes.php (hunk 7 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer/SmartQuotes.php:20]: Okay, I will do that once we agree with [my mentions here](https://github.com/jolicode/JoliTypo/pull/84#discussion_r1056212881).

File: src/JoliTypo/Fixer.php (hunk 5 above)
Comment (lyrixx) [reply src/JoliTypo/Fixer.php:49]: okay 👍🏼

File: src/JoliTypo/StateBag.php (hunk 11 above)
Comment (AmraniCh) [reply src/JoliTypo/StateBag.php]: Okey noted, I will consider to do this in the next PR.

File: src/JoliTypo/Fixer.php (hunk 5 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer.php]: ## LAST CHANGES UNTIL THE [COMMIT](https://github.com/jolicode/JoliTypo/pull/84/commits/19e324b8ad8d2391f02adeeff0f056c1f3206446)

* Removed all fully qualified class names and use namespaces imports instead.
* Removed all type declaration from public and protected properties and methods in the following classes: `FixerInterface` and `LocaleAwareFixerInterface` and all its implementing classes (including the `OkFixer` [see](https://github.com/jolicode/JoliTypo/pull/84/commits/f98efb97dbf616cd3749be3366e489f7c5f2c38a#diff-06f84166d2ab33296b16c55d86ec7430a4c6bbb2d8a10b817befba785eb97ef3)), `StateBag` and `Fixer\BaseOpenClosePair`.
* Added missing phpdoc return tags for verious methods in different classes.

**Comment by AmraniCh** [comment]
## LAST CHANGES UNTIL THE [COMMIT](https://github.com/jolicode/JoliTypo/pull/84/commits/19e324b8ad8d2391f02adeeff0f056c1f3206446)

* Removed all fully qualified class names and use namespaces imports instead.
* Removed all type declaration from public and protected properties and methods in the following classes: `FixerInterface` and `LocaleAwareFixerInterface` and all its implementing classes (including the `OkFixer` [see](https://github.com/jolicode/JoliTypo/pull/84/commits/f98efb97dbf616cd3749be3366e489f7c5f2c38a#diff-06f84166d2ab33296b16c55d86ec7430a4c6bbb2d8a10b817befba785eb97ef3)), `StateBag` and `Fixer\BaseOpenClosePair`.
* Added missing phpdoc return tags for verious methods in different classes.

File: src/JoliTypo/Fixer.php (hunk 5 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer.php]: done! see my last comment below for a general view of changes.

File: src/JoliTypo/Fixer.php [hunk 12]
```diff
@@ -97,10 +95,10 @@ public function fix($content)
 
     /**
      * @param string $content Basic content to fix
-     *
+     *
```
Comment (lyrixx) [inline src/JoliTypo/Fixer.php:98]: Oups :)

File: src/JoliTypo/Fixer.php [hunk 13]
```diff
@@ -111,35 +109,29 @@ public function fixString($content)
 
     /**
      * Change the list of rules for a given locale.
-     *
-     * @param array $rules Array of Fixer
-     *
-     * @throws Exception\BadRuleSetException
+     *
```
Comment (lyrixx) [inline src/JoliTypo/Fixer.php:112]: Can you remove all trailing space? Usually, there is a configuration option in your IDE.

**Review by lyrixx — COMMENTED** [review]
Yeah! This really great.
We are almost good!
Thanks for the hard work.

File: src/JoliTypo/Fixer/SmartQuotes.php (hunk 7 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer/SmartQuotes.php]: Done in [commit](https://github.com/jolicode/JoliTypo/pull/84/commits/f98efb97dbf616cd3749be3366e489f7c5f2c38a).

File: src/JoliTypo/Fixer.php (hunk 12 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer.php]: Done, thanks for the notice.

File: src/JoliTypo/Fixer.php (hunk 13 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer.php]: Done!

**Review by lyrixx — APPROVED** [review]
LGTM 👍🏼

File: src/JoliTypo/Fixer/Hyphen.php [hunk 14]
```diff
@@ -79,9 +82,9 @@ protected function setOptions()
     /**
      * Transform fr_FR to fr to fit the list of supported locales.
      *
-     * @return mixed
+     * @return void
```
Comment (damienalexandre) [inline src/JoliTypo/Fixer/Hyphen.php:85]: I think this method returns a string.

**Review by damienalexandre — APPROVED** [review]
Except for this one comment, everything looks perfect 👍 thanks a lot for this great work and giving your time to make JoliTypo better 😘

File: src/JoliTypo/Fixer/Hyphen.php (hunk 14 above)
Comment (AmraniCh) [reply src/JoliTypo/Fixer/Hyphen.php]: Emm you're right, how I missed this, however fixed and changed the return type to `string`.

**Comment by damienalexandre** [comment]
All good, thanks a lot!

**Comment by AmraniCh** [comment]
Thanks to you guys for this project. I will create other PRs as we have discussed before to add the other changes.
